#162685662 Enable Social login

tomuhenry opened the pull request:
**What does this PR do?**
This PR delivers a feature that enables social login via Google, Facebook, and Twitter

**Description of Task to be completed?**
The application should be able to allow social login through Facebook, Twitter, and Google. If the user isn't registered, the application should register the user, otherwise, it logs the user in.

**How should this be manually tested?**
- Add the following into the `.env` file
```
TWITTER_CONSUMER_KEY = annEeggI3NfH8xqWDTPU2O0Ve
TWITTER_CONSUMER_SECRET = w5tVQ6AnWluKdYUJo8aJrVurofat0Wl7qIwePvgqG3Sb57YtBg
```
- Install the requirements by `pip install -r requirements.txt`

**To test facebook Login**
- Create a developer account on facebook [Here](https://developers.facebook.com/) then create a [Facebook App](https://developers.facebook.com/apps) with your account 
- Got to the App's left menu and go to _Roles_, select _Test Users_ and add a test user. Select Edit and change permissions_, add _"email"_ and `update`
- Go to `edit` again and _get an access token for the user_
<img width="1114" alt="screen shot 2019-01-21 at 11 57 50 am" src="https://user-images.githubusercontent.com/40206038/51466562-9042f180-1d7b-11e9-9d9b-bc48843c5694.png">
- Copy that token and insert it in postman as shown in the image.
Use the POST method with URL `http://localhost:8000/api/users/facebook/`
<img width="1133" alt="screen shot 2019-01-21 at 12 03 04 pm" src="https://user-images.githubusercontent.com/40206038/51466581-98029600-1d7b-11e9-9ec0-2cd9e8126c9d.png">
- You can also go to [Facebook GraphApi Explorer](https://developers.facebook.com/tools/explorer/v2/) to get an access token for your self to test.
<img width="741" alt="screen shot 2019-01-21 at 11 56 15 am" src="https://user-images.githubusercontent.com/40206038/51466641-ba94af00-1d7b-11e9-9050-0a634bee1984.png">

**To Test Google Login**
- Go to [**Google Playground**](https://developers.google.com/oauthplayground/), select Google+ API, select all options and _Authorize APIs_. Sign in, Allow, and _Exchange Authorization Code for Tokens_ then copy the _**id_token**_ into postman as shown.
Use the POST method with URL `http://localhost:8000/api/users/google/`
<img width="935" alt="screen shot 2019-01-18 at 9 50 48 am" src="https://user-images.githubusercontent.com/40206038/51520873-ec6a4c00-1e35-11e9-8b1a-a4e3d2ee16b7.png">
<img width="1425" alt="screen shot 2019-01-21 at 12 01 11 pm" src="https://user-images.githubusercontent.com/40206038/51520896-fdb35880-1e35-11e9-866e-1b0af0a5e7f7.png">
<img width="1118" alt="screen shot 2019-01-21 at 12 02 04 pm" src="https://user-images.githubusercontent.com/40206038/51520916-0c9a0b00-1e36-11e9-9e47-ed4b0c82a402.png">

**To Test Twitter Login**
- Got to [Twitter Developer Page](https://developer.twitter.com) and log in. Create an **App** and got to the App details.
- Go to _Permissions_ and authorize access to the email address. 
- Go to _Keys and tokens_ and generate access token and token secret
- Copy the _Access token & access token secret_ to postman as shown and send.
Use the POST method with URL `http://localhost:8000/api/users/twitter/`
<img width="796" alt="screen shot 2019-01-22 at 11 42 49 am" src="https://user-images.githubusercontent.com/40206038/51523006-b62fcb00-1e3b-11e9-90ef-12c991b28b30.png">
<img width="878" alt="screen shot 2019-01-22 at 11 43 37 am" src="https://user-images.githubusercontent.com/40206038/51523021-bdef6f80-1e3b-11e9-9515-3975a3dcd2be.png">
<img width="844" alt="screen shot 2019-01-22 at 11 43 56 am" src="https://user-images.githubusercontent.com/40206038/51523026-c182f680-1e3b-11e9-8b95-afdb8d4b3ec0.png">
<img width="1087" alt="screen shot 2019-01-22 at 11 44 17 am" src="https://user-images.githubusercontent.com/40206038/51523035-cb0c5e80-1e3b-11e9-9d95-846c83a9ffb0.png">

**What are the relevant Pivotal Tracker stories?**
[#162685662](https://www.pivotaltracker.com/story/show/162685662)

archbishop12 commented on authors/apps/authentication/tests/test_social_auth.py:
```diff
@@ -0,0 +1,36 @@
+from rest_framework.views import status
+from django.urls import reverse
+
+class TestSocialLogin(APITestCase):
```
@tomuhenry Read about mocking/stubbing to write unit tests for the social login functionality.

Commit:
feature(Social Login):Enable Social login via Google, Facebook and Twitter

- Add serializer for social authentication
- Add url for the social login
- Create views for facebook, twitter, and google login
- Add social-auth to requirements files

[Maintains #162685662]

archbishop12 commented on authors/apps/authentication/tests/data.py:
```diff
@@ -152,3 +154,23 @@
+
+# Test Data for Social Login
+
+wrong_fb_token = {
```
@tomuhenry I don't think these token are required here.

tomuhenry replied:
Most of that is dummy data... It would probably not matter what you input as a token. It just needs some kind of well-formated input. Fixing it now

archbishop12 commented on authors/settings.py:
```diff
@@ -189,3 +193,6 @@
 EMAIL_HOST_USER = config('EMAIL_HOST_USER')
 EMAIL_HOST_PASSWORD = config('EMAIL_HOST_PASSWORD')
+
+TWITTER_CONSUMER_KEY = config('TWITTER_CONSUMER_KEY')
```
@tomuhenry Add these the `app.json`. Sync with @myrdstom to help with this. They should  be added to the `heroku environment variables` and `.env.example` file .

fahadmak reviewed (commented):
On google social endpoint try to catch errors for already registered users

moseskamira reviewed (approved):
LGTM, Basing on the Zoom Demo for the Functionality of this PR held Yesterday

archbishop12 reviewed (approved):
LGTM @tomuhenry i am merging this.
